pkg/agent/wireguard/client_linux.go

```diff
@@ -0,0 +1,240 @@
+		ListenPort:   &client.wireGuardConfig.WireGuardPort,
+		ReplacePeers: false,
+	}
+	patch, _ := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": map[string]string{
+				types.NodeWireGuardPublicKey: client.privateKey.PublicKey().String(),
+			},
+		},
+	})
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		_, err := client.k8sClient.CoreV1().Nodes().Patch(context.TODO(), client.nodeName, apitypes.MergePatchType, patch, metav1.PatchOptions{})
+		return err
+	}); err != nil {
+		return err
+	}
+
+	return client.wgClient.ConfigureDevice(client.wireGuardConfig.Name, cfg)
+}
+
+func (client *client) RemoveStalePeers(existingPeerPublickeys []string) error {
+	wgdev, err := client.wgClient.Device(client.wireGuardConfig.Name)
+	if err != nil {
+		return err
+	}
+	restoredPeerPublicKeys := make(map[wgtypes.Key]struct{})
+	for _, peer := range wgdev.Peers {
+		restoredPeerPublicKeys[peer.PublicKey] = struct{}{}
+	}
+	for _, e := range existingPeerPublickeys {
+		pubKey, err := wgtypes.ParseKey(e)
+		if err != nil {
+			klog.ErrorS(err, "Parse WireGuard public error")
+			continue
+		}
+		delete(restoredPeerPublicKeys, pubKey)
+	}
+	for k := range restoredPeerPublicKeys {
+		if err := client.deletePeerByPublicKey(k); err != nil {
+			klog.ErrorS(err, "Delete WireGuard peer error")
+		}
+	}
+	return nil
+}
+
+func (client *client) UpdatePeer(nodeName, publicKeyString string, peerNodeIP net.IP, podCIDRs []*net.IPNet) error {
+	pubKey, err := wgtypes.ParseKey(publicKeyString)
+	if err != nil {
+		return err
+	}
+	var allowedIPs []net.IPNet
+	peerNodeIPv4, peerNodeIPv6 := peerNodeIP.To4(), peerNodeIP.To16()
+
+	if peerNodeIPv4 != nil {
+		allowedIPs = append(allowedIPs, net.IPNet{
+			IP:   peerNodeIPv4,
+			Mask: net.CIDRMask(net.IPv4len*8, net.IPv4len*8),
+		})
+	} else if peerNodeIPv6 != nil {
+		allowedIPs = append(allowedIPs, net.IPNet{
+			IP:   peerNodeIPv6,
+			Mask: net.CIDRMask(net.IPv6len*8, net.IPv6len*8),
+		})
+	} else {
+		return fmt.Errorf("peer Node IP is not valid: %s", peerNodeIP.String())
+	}
+
+	for _, cidr := range podCIDRs {
+		allowedIPs = append(allowedIPs, *cidr)
+	}
+
+	if key, exist := client.peerPublicKeyByNodeName.Load(nodeName); exist {
+		cachedPeerPubKey := key.(wgtypes.Key)
+		if cachedPeerPubKey != pubKey {
+			klog.InfoS("WireGuard peer public key updated", "nodeName", nodeName, "publicKey", publicKeyString)
+			// delete old peer by public key.
+			if err := client.deletePeerByPublicKey(cachedPeerPubKey); err != nil {
+				return err
+			}
+		}
+	}
+	var peerNodeIPstring string
+	if peerNodeIPv4 != nil {
+		peerNodeIPstring = peerNodeIPv4.String()
+	} else {
+		peerNodeIPstring = peerNodeIPv6.String()
+	}
+	endpoint := net.JoinHostPort(peerNodeIPstring, strconv.Itoa(client.wireGuardConfig.WireGuardPort))
+	endpointUDP, err := net.ResolveUDPAddr("udp", endpoint)
+	if err != nil {
+		return err
+	}
+	client.peerPublicKeyByNodeName.Store(nodeName, pubKey)
```

Comment:
Not sure if deleting non-existing public key will have problem. If yes, it should be cached after ConfigureDevice succeeds.

Reply:
I have checked for this and it will not return errors for this case.